PullRequests for a product

iusyk opened the pull request:
Usage format: --prod_pulls "pull_number1[, pull_number2]"
for example,  to add the pulls 181 and 241 into the build,
the following has to be added to command line --prod_pulls "181, 241"

Known issues:
  1. In case of merge conflict(s) processing is terminated,
     the status of the repository is unpredictable

Implementation details:
- PyGithub library is used to work with GitHub
- build_prod.py is modified with the following functions
  1. verify_pulls: new function, verify if required PR exists
  2. get_pulls_for: new function, get the PRs for the certain repository
  3. handle_pulls_input: new function, transfer required PRs string into integers
  4. apply_pulls: new function, apply PRs to the repository
  5. process_pulls: new function, start the processing of the pull requests
  6. build_init: updated to deploy the remote, switch repository
     to the required branch, or setup the 'master' branch if branch
     is not defined and call process_pulls
- build_conf.py, support of the command line option '--prod_pulls' was added
  functions:
  get_prod_pulls: get the list of the PRs for the build

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>

iusyk commented:
@andr2000 @lorc hi all, please, take a look at this. Not ideal, but I think this is a point to discuss (or approve)
@andr2000 please, give me the permissions to assign the reviewers

andr2000 commented on build_prod.py:
```diff
@@ -51,6 +51,50 @@ def bash_run_command(cmd):
+# block to work with pull requests
+
+def verify_pulls(existed, proposed, uri):
+    print('Verifuing of the proposed pulls: {}'.format(proposed))
```
s/Verifu/Verify

andr2000 commented on build_prod.py:
```diff
@@ -51,6 +51,50 @@ def bash_run_command(cmd):
+    return list(map(int, inputstr.split(",")))
+
+
+def apply_pulls(dir, proposed):
```
dir is not used?

andr2000 commented on build_conf.py:
```diff
@@ -208,6 +211,10 @@ def __parse_args(self):
                             help="Use product's manifest from the branch specified")
+        parser.add_argument('--prod_pulls',
+                            dest='prod_pulls', required=False, default=False,
+                            help="Use to specify the list of the pull requests")
```
"...applied to the product's meta layer" so one is not confused with the impression we can apply pulls to the components

andr2000 commented on build_prod.py:
```diff
@@ -51,6 +51,50 @@ def bash_run_command(cmd):
+
+
+def get_pulls_for(uri):
+    print('!!!! pulls for {}'.format(uri))
```
do you need "!!!!" here?

andr2000 commented on build_prod.py:
```diff
@@ -51,6 +51,50 @@ def bash_run_command(cmd):
+
+
+def process_pulls(dir, uri, proposed):
+    proposed= handle_pulls_input(proposed)
```
s/= /=
Don't think you need an extra space there. Please change here and elsewhere

andr2000 commented on build_prod.py:
```diff
@@ -51,6 +51,50 @@ def bash_run_command(cmd):
+    res = []
+    for pr in pulls:
+        res.append(pr.number)
+    print(res)
```
Is this for debug only or how does it look like printed? Could you please give a sample output?

iusyk replied:
> Is this for debug only or how does it look like printed? Could you please give a sample output?

fixed

andr2000 commented on build_prod.py:
```diff
@@ -210,9 +254,25 @@ def add_meta_layers(cfg):
     repo_init(uri, branch, xml_base_name)
     repo_sync()
+    
+    c_dir = os.getcwd()
```
Could you please add a comment explaining what the magic here is

iusyk replied:
> Could you please add a comment explaining what the magic here is

added

andr2000 commented on build_prod.py:
```diff
@@ -210,9 +253,32 @@ def add_meta_layers(cfg):
+    os.chdir('meta-xt-{}'.format(xml_base_name))
+    # Get url of prod-layer to find the pull requests
+    # The name of repository is known but organization - not
+    # The organization can be extracted from theurl of layer 'meta-xt-products'
```
s/theurl/the url

andr2000 replied:
We hardcode github, products.git etc... So, may we also have "xen-troops" hardcoded as well?
Pull requests are mostly for development which happens on xen-troops.
Or I am missing the point and a developer can do this using his own github account to do the same?

iusyk replied:
I think, we should have a chance to work from own forks and have own meta-xt-products. Now (at least i assume) everybody can setup own environment,  BUT, i think we can improve the source code to work with the configuration from manifests.

iusyk replied:
fixed

rshym commented on README:
```diff
@@ -54,6 +54,8 @@ Optional arguments
 --parallel-build      - allow parallel build of domains.
                         This is useful for building on powerful machines or
                         build servers.
+--prod_pulls	      - allow define the list of the pull requests applied to 
```
May be it's better to use white spaces instead of tabs.
And for next line also.

iusyk replied:
fixed

rshym commented:
Also, I want to add that it could be good to reformat commit message to make it more easier to read.
Let me suggest following just as proposal and example of possible formatting.
I have changed usage of [] for optional argument (',' can be used only if pull_number2 is present) and rephrased some lines.

Just in case - it's only proposal.

```
Usage format: --prod_pulls "pull_number1[, pull_number2]"
for example,  to add the pulls 181 and 241 into the build,
the following has to be added to command line --prod_pulls "181, 241"

Known issues:
  1. In case of merge conflict(s) processing is terminated,
     the status of the repository is unpredictable

Implementation details:
- PyGithub library is used to work with GitHub
- build_prod.py is modified with the following functions
  1. verify_pulls: new function, verify if required PR exists
  2. get_pulls_for: new function, get the PRs for the certain repository
  3. handle_pulls_input: new function, transfer required PRs string into integers
  4. apply_pulls: new function, apply PRs to the repository
  5. process_pulls: new function, start the processing of the pull requests
  6. build_init: updated to deploy the remote, switch repository
     to the required branch, or setup the 'master' branch if branch
     is not defined and call process_pulls
- build_conf.py, support of the command line option '--prod_pulls' was added
  functions:
  get_prod_pulls: get the list of the PRs for the build

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>
```

iusyk commented:
> Also, I want to add that it could be good to reformat commit message to make it more easier to read.
> Let me suggest following just as proposal and example of possible formatting.
> I have changed usage of [] for optional argument (',' can be used only if pull_number2 is present) and rephrased some lines.
> 
> Just in case - it's only proposal.
> 
> ```
> Usage format: --prod_pulls "pull_number1[, pull_number2]"
> for example,  to add the pulls 181 and 241 into the build,
> the following has to be added to command line --prod_pulls "181, 241"
> 
> Known issues:
>   1. In case of merge conflict(s) processing is terminated,
>      the status of the repository is unpredictable
> 
> Implementation details:
> - PyGithub library is used to work with GitHub
> - build_prod.py is modified with the following functions
>   1. verify_pulls: new function, verify if required PR exists
>   2. get_pulls_for: new function, get the PRs for the certain repository
>   3. handle_pulls_input: new function, transfer required PRs string into integers
>   4. apply_pulls: new function, apply PRs to the repository
>   5. process_pulls: new function, start the processing of the pull requests
>   6. build_init: updated to deploy the remote, switch repository
>      to the required branch, or setup the 'master' branch if branch
>      is not defined and call process_pulls
> - build_conf.py, support of the command line option '--prod_pulls' was added
>   functions:
>   get_prod_pulls: get the list of the PRs for the build
> 
> Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>
> ```

changed, please check

rshym commented:
As I see, you changed text of very first comment for this pull request. And my proposal was about text of commit message itself. May be I'm wrong, but I see that commit message is same, but comment is changed.
In any case it was just proposal, and I'm ok with any option.

rshym commented on build_prod.py:
```diff
@@ -210,9 +253,32 @@ def add_meta_layers(cfg):
+    url = re.search('https://github.com/(.+?)products.git', uri).group(1)
+    url += xml_base_name
+    repo = git.Repo('./')
+    repo.create_remote('origin', url='git://github.com/{}'.format(url))
```
Let me suggest to change name of remote from 'origin' to something not so widely used. For example 'pull_req_source'.
Reason for that change is that we have distributed configuration across few projects, and we specify name of remote in product manifest in meta-xt-products. It will be strange to specify to not use 'origin' as default remote name because it may overlap with our build script. And there is no guarantee that some one will set default name of remote to 'origin'. So it's better to avoid usage of such common name.

iusyk replied:
changed

iusyk commented:
> As I see, you changed text of very first comment for this pull request. And my proposal was about text of commit message itself. May be I'm wrong, but I see that commit message is same, but comment is changed.
> In any case it was just proposal, and I'm ok with any option.

changed

rshym commented:
LGTM

andr2000 commented on build_prod.py:
```diff
@@ -210,11 +253,32 @@ def add_meta_layers(cfg):
+    os.chdir('meta-xt-{}'.format(xml_base_name))
+    # Get url of prod-layer to find the pull requests
+    # The name of repository is known but organization - not
+    # The organization can be extracted from theurl of layer 'meta-xt-products'
```
s/theurl/the url

andr2000 replied:
what if we do not have meta-xt-products?

andr2000 commented on build_prod.py:
```diff
@@ -210,11 +253,32 @@ def add_meta_layers(cfg):
+    # The organization can be extracted from theurl of layer 'meta-xt-products'
+    # Assumption:the layers  products and prod belong 
+    # to the same organization    
+    url = "xen-troops/meta-xt-" + xml_base_name
```
why `xen-troops` here? what if other organization is in place?

iusyk replied:
at this moment - I do not have a solution in scope of the existed command line
but we can add additional command line parameter to process it.
does it work ?

andr2000 commented on build_prod.py:
```diff
@@ -210,11 +253,32 @@ def add_meta_layers(cfg):
+    os.chdir(c_dir)
     # create build dir and make initial setup
-    yocto_run_command('')
+    #yocto_run_command('')
```
is it by intention?
